src/FSharp.Data.GraphQL.Shared/Validation.fs

```diff
@@ -1,60 +1,1057 @@
+                    match schemaInfo.TryGetTypeByRef(tref) with
+                    | Some itype ->
+                        let fieldMap = itype.InputFields |> Option.defaultValue [||] |> Array.map (fun x -> x.Name, x.Type) |> List.ofArray |> getFieldMap Map.empty
+                        let acc =
+                            fieldMap |> Seq.fold (fun acc kvp -> 
+                                if kvp.Value.Kind = TypeKind.NON_NULL && not (props.ContainsKey(kvp.Key))
+                                then acc @ Error.AsResult (sprintf "Can not coerce argument '%s'. Argument definition '%s' have a required field '%s', but that field does not exist in the literal value for the argument." argName tref.Name.Value kvp.Key, selection.Path)
+                                else acc) acc
+                        props |> Seq.fold (fun acc kvp ->
+                            match fieldMap.TryFind(kvp.Key) with
+                            | Some fieldTypeRef -> checkIsCoercible fieldTypeRef kvp.Key kvp.Value
+                            | None -> acc @ Error.AsResult(sprintf "Can not coerce argument '%s'. The field '%s' is not a valid field in the argument definition." argName kvp.Key, selection.Path)) acc
+                    | None -> canNotCoerce
+                | _ -> canNotCoerce
+            | Variable varName ->
+                let variableDefinition = 
+                    variables
+                    |> Option.defaultValue [] 
+                    |> List.tryFind (fun v -> v.VariableName = varName)
+                    |> Option.map (fun v -> v, schemaInfo.TryGetInputType(v.Type))
+                match variableDefinition with
+                | Some (vdef, Some vtype) when vdef.DefaultValue.IsSome -> checkIsCoercible vtype argName vdef.DefaultValue.Value
+                | Some (vdef, None) when vdef.DefaultValue.IsSome -> canNotCoerce
+                | _ -> acc
+        selection.Field.Arguments
+        |> List.fold (fun acc arg ->
+            let argumentTypeRef = selection.InputValues |> Option.defaultValue [||] |> Array.tryFind (fun x -> x.Name = arg.Name) |> Option.map (fun x -> x.Type)
+            match argumentTypeRef with
+            | Some argumentTypeRef -> acc @ (checkIsCoercible argumentTypeRef arg.Name arg.Value)
+            | None -> acc) acc
+
+    let internal validateInputValues (ctx : ValidationContext) =
+        ctx.Definitions
+        |> List.choose (function
+            | OperationDefinitionInfo odef -> Some (Some odef.Definition.VariableDefinitions, odef.SelectionSet)
+            | FragmentDefinitionInfo fdef -> Some (None, fdef.SelectionSet))
+        |> List.fold (fun acc (vars, selectionSet) -> selectionSet |> List.fold (fun acc selection -> checkInputValue acc ctx.Schema vars selection) acc) Success
+
+    let rec private getDistinctDirectiveNamesInSelection (path : Path) (selection : Selection) : (Path * string list) list =
```

Comment:
Instead of using a list along with `distinct`, could we use a set instead?

Reply:
Done.